File: 1293.shortest-path-in-a-grid-with-obstacles-elimination.cs
/*
 * @lc app=leetcode id=1293 lang=csharp
 *
 * [1293] Shortest Path in a Grid with Obstacles Elimination
 */

// @lc code=start
public class Solution {
    public int ShortestPath(int[][] grid, int k) {
        int m = grid.Length;
        int n = grid[0].Length;
        int steps = 0;
        Queue<int[]> queue = new Queue<int[]>();
        Dictionary<Tuple<int,int>, int> visited = new Dictionary<Tuple<int,int>, int>();
        queue.Enqueue(new int[3]{0, 0, 0});
        visited[new Tuple<int,int>(0,0)] = 0;
        while (queue.Count > 0)
        {
            int l = queue.Count;
            for (int i = 0; i < l; i++)
            {
                int[] item = queue.Dequeue();
                //Console.WriteLine($"{item[0]}, {item[1]}, {item[2]}");
                if (item[0] == m-1 && item[1] == n-1)
                {
                    return steps;
                }
                int[][] neigs = new int[4][];
                neigs[0] = new int[]{item[0]-1, item[1]};
                neigs[1] = new int[]{item[0]+1, item[1]};
                neigs[2] = new int[]{item[0], item[1]-1};
                neigs[3] = new int[]{item[0], item[1]+1};
                foreach (int[] nxt in neigs)
                {
                    int xn = nxt[0];
                    int yn = nxt[1];
                    if (xn >= 0 && xn < m && yn >= 0 && yn < n)
                    {
                        int state = item[2];
                        if (grid[xn][yn] == 1)
                        {
                            state += 1;
                        }
                        Tuple<int,int> newKey = new Tuple<int,int>(xn,yn);
                        if (state <= k && (!visited.ContainsKey(newKey) || visited[newKey] > state))
                        {
                            queue.Enqueue(new int[3]{xn, yn, state});
                            visited[newKey] = state;
                        }
                    }
                }
            }
            steps += 1;
        }
        return -1;
    }
}

// @lc code=end

File: 1293.shortest-path-in-a-grid-with-obstacles-elimination.py
#
# @lc app=leetcode id=1293 lang=python3
#
# [1293] Shortest Path in a Grid with Obstacles Elimination
#

# @lc code=start
class Solution:
    # BFS will do the job
    # We BFS from the upper left corner.
    # state = (x,y,obstacles removed)
    # When we see an obstacle, we add one to obstacle
    # We also keep a hash table mapping coordinates to minimum obstacles to remove to reach the point
    # See if we can arrive at the the lower right
    
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        m, n, steps, dq, visited = len(grid), len(grid[0]), 0, deque(), {}
        dq.append((0,0,0))
        visited[(0,0)] = 0
        while dq:
            for i in range(len(dq)):
                x, y, removed = dq.popleft()
                if x == m-1 and y == n-1:
                    return steps
                neigs = [(x-1, y), (x+1,y), (x,y-1),(x,y+1)]
                for xn, yn in neigs:
                    if xn >= 0 and yn >= 0 and xn < m and yn < n:
                        state = removed
                        if grid[xn][yn] == 1:
                            state += 1
                        if state <= k and ((xn, yn) not in visited or visited[(xn, yn)] > state):
                            dq.append((xn, yn, state))
                            visited[(xn,yn)] = state
            steps += 1
        return -1
# @lc code=end

